### PR TITLE
Refactor tests for union sorting properties, add tests for unions and constants

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Part of https://github.com/apache/datafusion/issues/12446, broken out of https://github.com/apache/datafusion/pull/12562 for easy review

## Rationale for this change

In order to write tests for https://github.com/apache/datafusion/issues/12446 I found I needed to be able to set SortOptions and constants more easily.

## What changes are included in this PR?
1. Refactor the tests for union property calculation into a structure.
2. Add `#[ignored]` tests for https://github.com/apache/datafusion/issues/12446

## Are these changes tested?
Only tests

## Are there any user-facing changes?
No functional changes

### CODE REVIEW

File: datafusion/physical-expr/src/equivalence/ordering.rs [hunk 1]
```diff
@@ -36,15 +36,15 @@ use arrow_schema::SortOptions;
 ///
 /// Here, both `vec![a ASC, b ASC]` and `vec![c DESC, d ASC]` describe the table
 /// ordering. In this case, we say that these orderings are equivalent.
-#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Default)]
```
Comment (alamb) [inline datafusion/physical-expr/src/equivalence/ordering.rs:39]: this is a small drive by cleanup -- I can move it to its own PR if preferred

File: datafusion/physical-expr/src/equivalence/properties.rs [hunk 2]
```diff
@@ -2708,379 +2708,428 @@ mod tests {
         ))
     }
 
-    #[tokio::test]
-    async fn test_union_equivalence_properties_multi_children() -> Result<()> {
-        let schema = create_test_schema()?;
+    #[test]
+    fn test_union_equivalence_properties_multi_children_1() {
```
Comment (alamb) [inline datafusion/physical-expr/src/equivalence/properties.rs:2712]: The individual functions corresponds to TEST CASE N. 

I think they are easier to understand, are more concise and run in parallel (and I can run individual tests)

File: datafusion/physical-expr/src/equivalence/properties.rs [hunk 3]
```diff
@@ -2708,379 +2708,428 @@ mod tests {
...
-                .map(|ordering| convert_to_sort_exprs(ordering))
-                .collect::<Vec<_>>();
-            let second_constants = second_child_constants
-                .iter()
-                .map(|expr| ConstExpr::new(Arc::clone(expr)))
-                .collect::<Vec<_>>();
-            let mut rhs = EquivalenceProperties::new(Arc::clone(second_schema));
-            rhs = rhs.with_constants(second_constants);
-            rhs.add_new_orderings(second_orderings);
+    #[test]
```
Comment (alamb) [inline datafusion/physical-expr/src/equivalence/properties.rs:2891]: these are new tests for https://github.com/apache/datafusion/issues/12446 that currently fail

File: datafusion/physical-expr/src/equivalence/properties.rs [hunk 4]
```diff
@@ -2708,379 +2708,428 @@ mod tests {
...
+                // Union orderings:
+                // [a, b, d] (c constant)
+                // [a, c, d] (b constant)
+                vec![vec!["a", "c", "b DESC", "d"], vec!["a", "b DESC", "c", "d"]],
+                vec![],
+            )
+            .run()
+    }
+
+    // TODO tests with multiple constants
```
Comment (alamb) [inline datafusion/physical-expr/src/equivalence/properties.rs:3008]: I will do this as part of https://github.com/apache/datafusion/issues/12446

File: datafusion/physical-expr/src/equivalence/properties.rs [hunk 5]
```diff
@@ -2708,379 +2708,428 @@ mod tests {
...
 
-            let union_expected_orderings = union_orderings
+    #[derive(Debug)]
+    struct UnionEquivalenceTest {
```
Comment (alamb) [inline datafusion/physical-expr/src/equivalence/properties.rs:3011]: I pulled all the test boilerplate into a struct that can be built / ordered more easily

File: datafusion/physical-expr/src/equivalence/properties.rs [hunk 6]
```diff
@@ -3091,21 +3140,63 @@ mod tests {
...
-        assert_eq!(lhs_constants.len(), rhs_constants.len(), "{}", err_msg);
         for rhs_constant in rhs_constants {
             assert!(
                 const_exprs_contains(lhs_constants, rhs_constant.expr()),
-                "{}",
-                err_msg
+                "{err_msg}\nlhs: {lhs}\nrhs: {rhs}"
             );
         }
+        assert_eq!(
```
Comment (alamb) [inline datafusion/physical-expr/src/equivalence/properties.rs:3149]: moved afterwards so if the contents are different we get the contents printed out first

File: datafusion/physical-expr/src/equivalence/properties.rs (hunk 2 above)
Comment (dmitrybugakov) [reply datafusion/physical-expr/src/equivalence/properties.rs:2712]: After the refactor, it looks great.

**Review by dmitrybugakov — APPROVED** [review]

File: datafusion/physical-expr/src/equivalence/ordering.rs [hunk 7]
```diff
@@ -15,13 +15,13 @@
...
-use std::hash::Hash;
-use std::sync::Arc;
-
 use crate::equivalence::add_offset_to_expr;
 use crate::{LexOrdering, PhysicalExpr, PhysicalSortExpr};
 use arrow_schema::SortOptions;
+use std::fmt::Display;
+use std::hash::Hash;
+use std::sync::Arc;
+use std::vec::IntoIter;
```
Comment (berkaysynnada) [inline datafusion/physical-expr/src/equivalence/ordering.rs:25]: I think we can move these up where they were :)

File: datafusion/physical-expr/src/equivalence/ordering.rs (hunk 1 above)
Comment (berkaysynnada) [reply datafusion/physical-expr/src/equivalence/ordering.rs:39]: I think there is no need to separate.

File: datafusion/physical-expr/src/equivalence/properties.rs [hunk 8]
```diff
@@ -2708,379 +2708,428 @@ mod tests {
...
+                expected_properties: None,
+            }
+        }
+
+        /// Add a union input with the specified orderings
+        ///
+        /// See [`Self::make_props`] for the format of the strings in `orderings`
+        fn with_child_sort(
+            mut self,
+            orderings: Vec<Vec<&'static str>>,
```
Comment (berkaysynnada) [inline datafusion/physical-expr/src/equivalence/properties.rs:3035]: I think we can remove these static lifetimes in the utils.

File: datafusion/physical-expr/src/equivalence/properties.rs (hunk 4 above)
Comment (berkaysynnada) [reply datafusion/physical-expr/src/equivalence/properties.rs:3008]: Even better to add some negative tests for the new fix -- having a few constants could lead many existing orderings to spread after the union. We should make sure we do not preserve the equivalences when we really cannot.

Comment (alamb) [reply datafusion/physical-expr/src/equivalence/properties.rs:3008]: That is an excellent idea -- I will do so as part of https://github.com/apache/datafusion/issues/12446

File: datafusion/physical-expr/src/equivalence/properties.rs (hunk 8 above)
Comment (alamb) [reply datafusion/physical-expr/src/equivalence/properties.rs]: done

**Comment by alamb** [comment]
Thank you again for the very speedy review @dmitrybugakov and @berkaysynnada

**Pull request merged**